File: killrvideo-commons/src/main/java/com/killrvideo/dse/utils/AsyncResultSetUtils.java
package com.killrvideo.dse.utils;

import com.datastax.oss.driver.api.core.cql.AsyncResultSet;
import com.datastax.oss.driver.api.core.cql.Row;
import com.datastax.oss.protocol.internal.util.Bytes;
import com.killrvideo.dse.dto.ResultListPage;

import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class AsyncResultSetUtils {
    private AsyncResultSetUtils() {
    }

    public static <T> ResultListPage<T> toResultListPage(AsyncResultSet rs, Function<Row, T> mapper) {
        List<T> list = getResultsOnCurrentPage(rs, mapper);
        Optional<String> nextPage = getPagingState(rs);
        return new ResultListPage<>(list, nextPage);
    }

    public static <T> List<T> getResultsOnCurrentPage(AsyncResultSet rs, Function<Row, T> mapper) {
        return StreamSupport.stream(rs.currentPage().spliterator(), false)
            .map(mapper).collect(Collectors.toList());
    }

    public static Optional<String> getPagingState(AsyncResultSet rs) {
        if (rs.hasMorePages() && rs.getExecutionInfo().getPagingState() != null) {
            return Optional.ofNullable(Bytes.toHexString(rs.getExecutionInfo().getPagingState()));
        } else {
            return Optional.empty();
        }
    }
}

File: killrvideo-commons/src/main/kotlin/com/killrvideo/dse/utils/AsyncResultSetUtils.kt
package com.killrvideo.dse.utils

import com.datastax.oss.driver.api.core.cql.AsyncResultSet
import com.datastax.oss.driver.api.core.cql.Row
import com.datastax.oss.protocol.internal.util.Bytes
import com.killrvideo.dse.dto.ResultListPage
import java.util.function.Function
import java.util.stream.Collectors
import java.util.stream.StreamSupport

object AsyncResultSetUtils {
    fun <T> toResultListPage(rs: AsyncResultSet, mapper: Function<Row, T>): ResultListPage<T> {
        val list = getResultsOnCurrentPage(rs, mapper)
        val nextPage = getPagingState(rs)
        return ResultListPage(list, nextPage)
    }

    private fun <T> getResultsOnCurrentPage(rs: AsyncResultSet, mapper: Function<Row, T>): List<T> {
        return StreamSupport.stream(rs.currentPage().spliterator(), false)
            .map(mapper).collect(Collectors.toList())
    }

    fun getPagingState(rs: AsyncResultSet): String? {
        return if (rs.hasMorePages() && rs.executionInfo.pagingState != null) {
            Bytes.toHexString(rs.executionInfo.pagingState)
        } else {
            null
        }
    }
}
